openmeter/app/stripe/entity/app/app_invoice.go

```diff
@@ -0,0 +1,426 @@
+	stripeCustomerData, err := a.StripeAppService.GetStripeCustomerData(ctx, appstripeentity.GetStripeCustomerDataInput{
+		AppID:      a.GetID(),
+		CustomerID: customerID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get stripe customer data: %w", err)
+	}
+
+	// This should never happen as invoices call ValidateInvoice before calling this method.
+	if stripeCustomerData.StripeDefaultPaymentMethodID == nil {
+		return nil, app.CustomerPreConditionError{
+			AppID:      a.GetID(),
+			AppType:    a.GetType(),
+			CustomerID: customerID,
+			Condition:  "default payment method cannot be null",
+		}
+	}
+
+	// Create the invoice in Stripe
+	stripeInvoice, err := client.Invoices.New(&stripe.InvoiceParams{
+		// FinalizeInvoice will advance the invoice
+		AutoAdvance:          stripe.Bool(false),
+		Currency:             stripe.String(string(invoice.Currency)),
+		Customer:             stripe.String(stripeCustomerData.StripeCustomerID),
+		DueDate:              lo.ToPtr(invoice.DueAt.Unix()),
+		DefaultPaymentMethod: stripe.String(*stripeCustomerData.StripeDefaultPaymentMethodID),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create invoice in stripe: %w", err)
+	}
+
+	// Return the result
+	result := &billing.UpsertInvoiceResult{}
+	result.SetExternalID(stripeInvoice.ID)
+	result.SetInvoiceNumber(stripeInvoice.Number)
+
+	// Add line items
+	stripeInvoiceLineParams := &stripe.InvoiceAddLinesParams{}
+
+	// Walk the tree
+	var queue []*billing.Line
+
+	// Feed the queue with the root lines
+	invoice.Lines.ForEach(func(lines []*billing.Line) {
+		queue = append(queue, lines...)
+	})
+
+	// We collect the line IDs to match them with the Stripe line items from the response
+	var lineIDs []string
+
+	for len(queue) > 0 {
+		line := queue[0]
+		queue = queue[1:]
+
+		// Add children to the queue
+		childrens := line.Children.OrEmpty()
+		for _, l := range childrens {
+			queue = append(queue, l)
+		}
+
+		// Only add line items for leaf nodes
+		if len(childrens) > 0 {
+			continue
+		}
+
+		period := &stripe.InvoiceAddLinesLinePeriodParams{
+			Start: lo.ToPtr(line.Period.Start.Unix()),
+			End:   lo.ToPtr(line.Period.End.Unix()),
+		}
+
+		// Add discounts
+		line.Discounts.ForEach(func(discounts []billing.LineDiscount) {
+			for _, discount := range discounts {
+				lineIDs = append(lineIDs, line.ID)
+
+				stripeInvoiceLineParams.Lines = append(stripeInvoiceLineParams.Lines, &stripe.InvoiceAddLinesLineParams{
+					Description: discount.Description,
+					Amount:      lo.ToPtr(-discount.Amount.GetFixed()),
+					Quantity:    lo.ToPtr(int64(1)),
+					Period:      period,
+				})
+			}
+		})
+
+		// Add line item
+		switch line.Type {
+		case billing.InvoiceLineTypeFee:
+			lineIDs = append(lineIDs, line.ID)
+
+			stripeInvoiceLineParams.Lines = append(stripeInvoiceLineParams.Lines, &stripe.InvoiceAddLinesLineParams{
+				Description: line.Description,
+				Amount:      lo.ToPtr(line.Totals.Amount.GetFixed()),
```

Comment:
This is incorrect, due to number representation.

GetFixed gets the fixed point representation of the number if available (I think it will use ~10 binary digits).

This only exists for optimized serialization of the alpacadecimal package.

The stripe API states that:
```
	// The integer amount in cents (or local equivalent) of the charge to be applied to the upcoming invoice. If you want to apply a credit to the customer's account, pass a negative amount.
	Amount *int64 `form:"amount"`
```

So here you should use the currencyx/gobl currency package to determine the subunits of the given currency and get an integer representation rounded to the base 10 decimal places.

```